docs/rfc/block-registration.md

```diff
@@ -0,0 +1,369 @@
+This RFC is intended to serve both as a specification and as documentation for the implementation of runtime-agnostic block type registration.
+
+## Requirements
+
+Behind any block type registration is some abstract concept of a unit of content. This content type can be described without consideration of any particular technology. In much the same way, we should be able to describe the core constructs of a block type in a way which can be interpreted in any runtime.
+
+In more practical terms, an implementation should fulfill requirements that...
+
+* A block type registration should be declarative and context-agnostic. Any runtime (PHP, JS, or other) should be able to interpret the basics of a block type (see "Block API" in the sections below) and should be able to fetch or retrieve the definitions of the context-specific implementation details. The following things should be made possible:
+  * Building REST APIs to fetch the available block types
+  * Building REST APIs to fetch block objects from posts
+* This API should be backward compatible with what we have at the moment.
+* It should be possible to statically analyze a block type in order to support advanced use-cases required by one of the [9 projects](https://make.wordpress.org/core/2018/12/08/9-priorities-for-2019/) for 2019 in Wordpress: "Building a WordPress.org directory for discovering blocks, and a way to seamlessly install them."
+* It should not require a build tool compilation step (e.g. Babel, Webpack) to author code which would be referenced in a block type definition.
+* There should allow the potential to dynamically load ("lazy-load") block types, or parts of block type definitions.
```

Comment:
- It means the editor should be able to be shown/loaded without enqueuing all the assets (scripts/styles) of all block types: It can load the titles/icons... synchronously but leave all the `edit/save/transforms`... for the moment they are used (inserted into the post content).

For example, a playlist block won't load mediaelement unless used, a syntax highlighter won't be loaded unless the code/html blocks are loaded...

Another aspect of the lazy-loading is the possibility to enqueue only the styles/scripts required by blocks used in the frontend.

It doesn't mean we're going to solve these right now, but it means, the registration API should allow it.

Reply:
JS and CSS resources wouldn't be loaded by the browser until they would be explicitly selected in UI. More in Webpack docs: https://webpack.js.org/guides/lazy-loading/

I'm sure we can also find easier to follow articles online.

Reply:
>     * It means the editor should be able to be shown/loaded without enqueuing all the assets (scripts/styles) of all block types: It can load the titles/icons... synchronously but leave all the `edit/save/transforms`... for the moment they are used (inserted into the post content).

Is it worth including this in the RFC? I think this is implied knowledge that would be better off as explicitly stated.

Reply:
Clarified in b8d0e1b as suggested.